TLM/TLM/Util/Extensions/NetSegmentExtensions.cs

```diff
@@ -283,63 +284,69 @@ public static class NetSegmentExtensions {
-                                ? NetInfo.Direction.Backward
-                                : NetInfo.Direction.Forward;
-                filterDir = inverted
-                                ? NetInfo.InvertDirection((NetInfo.Direction)filterDir)
-                                : filterDir;
-                sortDir = NetInfo.InvertDirection((NetInfo.Direction)filterDir);
-            } else if (inverted) {
-                sortDir = NetInfo.Direction.Backward;
-            }
+            NetInfo.Direction sortDir;
+
+            if (startNode.HasValue) {
+                filterDir = startNode.Value
+                    ? NetInfo.Direction.Backward
+                    : NetInfo.Direction.Forward;
 
-            if (reverse) {
-                sortDir = NetInfo.InvertDirection(sortDir);
+                if (inverted)
+                    filterDir = NetInfo.InvertDirection(filterDir.Value);
+
+                sortDir = NetInfo.InvertDirection(filterDir.Value);
+            } else {
+                sortDir = inverted
+                    ? NetInfo.Direction.Backward
+                    : NetInfo.Direction.Forward;
             }
 
             uint curLaneId = netSegment.m_lanes;
             byte laneIndex = 0;
 
             NetManager netManager = Singleton<NetManager>.instance;
 
-            while (laneIndex < segmentInfo.m_lanes.Length && curLaneId != 0u) {
+            while (laneIndex < maxLanes && curLaneId != 0u) {
                 NetInfo.Lane laneInfo = segmentInfo.m_lanes[laneIndex];
-                if ((laneTypeFilter == null ||
-                     (laneInfo.m_laneType & laneTypeFilter) != NetInfo.LaneType.None) &&
-                    (vehicleTypeFilter == null || (laneInfo.m_vehicleType & vehicleTypeFilter) !=
-                     VehicleInfo.VehicleType.None) &&
-                    (filterDir == null ||
-                     laneInfo.m_finalDirection == filterDir)) {
-                    laneList.Add(
-                        new LanePos(
-                            curLaneId,
-                            laneIndex,
-                            laneInfo.m_position,
-                            laneInfo.m_vehicleType,
-                            laneInfo.m_laneType));
+
+                if ((laneTypeFilter == null || (laneInfo.m_laneType & laneTypeFilter) != 0) &&
+                    (vehicleTypeFilter == null || (laneInfo.m_vehicleType & vehicleTypeFilter) != 0) &&
+                    (filterDir == null || laneInfo.m_finalDirection == filterDir)) {
+
+                    laneList.Add(new LanePos(
+                        curLaneId,
+                        laneIndex,
+                        laneInfo.m_position,
+                        laneInfo.m_vehicleType,
+                        laneInfo.m_laneType));
                 }
 
                 curLaneId = netManager.m_lanes.m_buffer[curLaneId].m_nextLane;
                 ++laneIndex;
             }
 
-            if (sort) {
+            laneList.TrimExcess();
```

Comment:
This will cause an additional gc, is it better to remove it at expense of using a little more RAM?

Reply:
doesn't matter.